### PR TITLE
UUID tplg support

### PR DESCRIPTION
**Pull request by keyonjie**
This is the series for UUID support -- tplg part, it should be used together with the FW part and kernel part.

### CODE REVIEW

**Comment by keyonjie** [comment]
the corresponding Firmware PR: https://github.com/thesofproject/sof/pull/2920
the corresponding Linux PR: https://github.com/thesofproject/linux/pull/2090

File: tools/topology/m4/muxdemux.m4 [hunk 1]
```diff
@@ -36,6 +36,15 @@ define(`N_MUXDEMUX', `MUXDEMUX'PIPELINE_ID`.'$1)
 
 dnl W_MUXDEMUX(name, mux/demux, format, periods_sink, periods_source, core, kcontrol_list)
 define(`W_MUXDEMUX',
+`SectionVendorTuples."'N_MUXDEMUX($1)`_tuples_uuid" {'
+`	tokens "sof_comp_tokens"'
+`	tuples."uuid" {'
+`		SOF_TKN_COMP_UUID'		"c4:b2:68:68:14:30:47:0e:a0:89:15:d1:c7:7f:85:1a"
+`	}'
```
Comment (lgirdwood) [inline tools/topology/m4/muxdemux.m4:43]: It's better here to have a uuid.m4 file that contains all UUIDs and is included by all topologies.
```
# uuid.m4 - set default UUIDs if NOT already set in topology.
# allows users to override the UUID if needed.
ifndef UUID_SRC
define UUID_SRC "blah blah blah"
endif

# and so on
```

File: tools/topology/m4/muxdemux.m4 [hunk 2]
```diff
@@ -36,6 +36,15 @@ define(`N_MUXDEMUX', `MUXDEMUX'PIPELINE_ID`.'$1)
 
 dnl W_MUXDEMUX(name, mux/demux, format, periods_sink, periods_source, core, kcontrol_list)
 define(`W_MUXDEMUX',
+`SectionVendorTuples."'N_MUXDEMUX($1)`_tuples_uuid" {'
+`	tokens "sof_comp_tokens"'
+`	tuples."uuid" {'
+`		SOF_TKN_COMP_UUID'		"c4:b2:68:68:14:30:47:0e:a0:89:15:d1:c7:7f:85:1a"
```
Comment (mmaka1) [inline tools/topology/m4/muxdemux.m4:42]: Could you document the format used for uuid in this token?
If this is just a byte array, is endianess correct? There are uint32, uint32, uint8.
Example from the fw source code:
```c
/* c4b26868-1430-470e-a089-15d1c77f851a */
DECLARE_SOF_UUID("demux", demux_uuid, 0xc4b26868, 0x1430, 0x470e,
		 0xa0, 0x89, 0x15, 0xd1, 0xc7, 0x7f, 0x85, 0x1a);
```
So if it is byte array and little-endian is assumed, it should rather be 68:68:b2:c4:30:14....

Comment (keyonjie) [reply tools/topology/m4/muxdemux.m4:42]: You are right, thanks for spotting this.

File: tools/topology/m4/muxdemux.m4 (hunk 1 above)
Comment (keyonjie) [reply tools/topology/m4/muxdemux.m4:43]: As discussed, will implement a m4 helper to generate this UUID byte array from the string copied from the firmware source directly.

**Comment by keyonjie** [comment]
@lgirdwood @mmaka1 I want to let this (adding UUID tokens to toplogies) go first, they are verified with the corresponding FW and kernel changes, but the topology changes can go first as they won't be used if driver and FW is not ready yet, no any regression with these changes applied.

**Comment by keyonjie** [comment]
Not all components are covered yet, there will be more commits added for them, e.g. asrc/buffer/dai/dcblock/eq/mixer/pcm/pga/src/tone...

**Comment by keyonjie** [comment]
> I guess there is nothing stopping this being merged after kernel token is updated ? i.e. it's not used so wont change anything ?

Right. We can merge this even before the kernel token is updated, which means we provide those tokens in tplg, but the kernel doesn't parse them so nothing will change to others.

**Comment by keyonjie** [comment]
> What's the status of the kernel PR for the new token ID ?

That is the third commit of the driver PR, https://github.com/thesofproject/linux/pull/2090/commits/04a7e7a3611c917716e7399f8a3e6f91d92e57d8

@kv2019i @plbossart I think we are all aligned on adding this token on both tplg and driver side?

**Comment by plbossart** [comment]
> > What's the status of the kernel PR for the new token ID ?
> 
> That is the third commit of the driver PR, [thesofproject/linux@04a7e7a](https://github.com/thesofproject/linux/commit/04a7e7a3611c917716e7399f8a3e6f91d92e57d8)
> 
> @kv2019i @plbossart I think we are all aligned on adding this token on both tplg and driver side?

I am not, there are 4 parallel threads (sof-developers, sof-docs, sof, linux), it's beyond my reviewer ability quite frankly to figure out what the big picture looks like.

**Comment by kv2019i** [comment]
@plbossart wrote:
> I am not, there are 4 parallel threads (sof-developers, sof-docs, sof, linux), it's beyond my reviewer ability quite frankly to figure out what the big picture looks like.

I suggest looking at the sof-docs:
https://github.com/thesofproject/sof-docs/pull/245

.. let's all align on that first. If we all approve, then reviewing the others should be easier as they should all comply with the sof-docs. Ok @plbossart @keyonjie @lgirdwood ?

**Comment by keyonjie** [comment]
> @plbossart wrote:
> 
> > I am not, there are 4 parallel threads (sof-developers, sof-docs, sof, linux), it's beyond my reviewer ability quite frankly to figure out what the big picture looks like.
> 
> I suggest looking at the sof-docs:
> [thesofproject/sof-docs#245](https://github.com/thesofproject/sof-docs/pull/245)
> 
> .. let's all align on that first. If we all approve, then reviewing the others should be easier as they should all comply with the sof-docs. Ok @plbossart @keyonjie @lgirdwood ?

Good point, ack.

**Comment by jajanusz** [comment]
@keyonjie pls fix ci errors

**Comment by keyonjie** [comment]
SOFCI TEST

**Comment by keyonjie** [comment]
> @keyonjie pls fix ci errors

I can build it locally with latest alsa-lib, looks the CI is using quite old alsa-lib? @xiulipan

**Comment by xiulipan** [comment]
@lgirdwood Can we do a update and rebase for the https://github.com/thesofproject/alsa-lib? It seems we need latest alsa-lib for this feature.

**Comment by lgirdwood** [comment]
@xiulipan please do.

**Comment by xiulipan** [comment]
@lgirdwood It seems I did not have permission to push branch to alsa-lib. I pushed to rebased branch to my repo https://github.com/xiulipan/alsa-lib
alsa-utils did not need rebase, a simple fast-forward update would be enough.

Please help to update the branch or give me push access for those 2 repos.

**Comment by lgirdwood** [comment]
@xiulipan you should now have access.

**Comment by lgirdwood** [comment]
@xiulipan UUID FW now merged.

**Comment by xiulipan** [comment]
SOFCI TEST

**Comment by xiulipan** [comment]
@lgirdwood @keyonjie The Docker image updated and I re-trigger tests. It seems we missed the testbench patch for UUID.

**Comment by marc-hb** [comment]
> It seems we missed the testbench patch for UUID.

So we have 3 topology parsers total, correct?
- alsa-lib https://www.alsa-project.org/wiki/ALSA_topology
- https://github.com/thesofproject/sof/tree/master/tools/tplg_parser
- https://github.com/thesofproject/sof-test/tree/master/tools

**Comment by lgirdwood** [comment]
> > It seems we missed the testbench patch for UUID.
> 
> So we have 3 topology parsers total, correct?
> 
> * alsa-lib https://www.alsa-project.org/wiki/ALSA_topology

This one is the compiler.

> * https://github.com/thesofproject/sof/tree/master/tools/tplg_parser
> * https://github.com/thesofproject/sof-test/tree/master/tools

Test should be using the tools version - please fix in test.

**Comment by lgirdwood** [comment]
> Test should be using the tools version - please fix in test.

To clarify, test should use SOF tools version, but my understanding is that this in in Python today in test so something will need to be worked out as it will be expensive to maintain both.

**Comment by lgirdwood** [comment]
@zrombel can you check CI (nothing is showing red) - I'm told by @keyonjie that the test should be passing.
@keyonjie does the CI needs updated for this test to pass now ?

**Comment by keyonjie** [comment]
> @zrombel can you check CI (nothing is showing red) - I'm told by @keyonjie that the test should be passing.
> @keyonjie does the CI needs updated for this test to pass now ?

I can't see the QB log, instead seeing all SUCCESS while it reports Tools failed. And for the alsa-lib version in CI, yes, let's align using latest version for building.

**Comment by keyonjie** [comment]
@slawblauciak could you help to take a look?

**Comment by keyonjie** [comment]
SOFCI TEST

**Comment by keyonjie** [comment]
@lgirdwood , @xiulipan told me that it is known issue for QB on TGL, so I think we are good to go with this.

**Comment by keyonjie** [comment]
@zrombel please make sure the QB server is using alsa-lib later than v1.2.1

**Comment by marc-hb** [comment]
> @zrombel please make sure the QB server is using alsa-lib later than v1.2.1

I have a branch that lets you build 1.2.3.2 .deb files for Ubuntu in two commands:
https://github.com/thesofproject/linux/issues/2211#issuecomment-656418209

Feedback appreciated there.

**Comment by keyonjie** [comment]
@zrombel forgot to mention, please upgrade both alsa-lib and alsa-utils to the same version v1.2.3.

**Comment by zrombel** [comment]
Process Team is working on updating QB machines with newest alsa-lib and utils. We were able to run one test build on updated machine on this PR and it'a PASS so this PR can be merged. Rest of the machines are still under update. QB should be fully ready at the end of the day.

**Comment by keyonjie** [comment]
Thanks @zrombel . @lgirdwood Let's merge this and then kernel PR will go together.

**Comment by keyonjie** [comment]
SOFCI TEST

**Comment by zrombel** [comment]
QB CI is now fully updated.

**Pull request merged**